src/libNetwork/DataSender.cpp

```diff
@@ -0,0 +1,257 @@
+  unsigned int shard_groups_count = shards.size() / num_clusters;
+  if ((shards.size() % num_clusters) > 0) {
+    shard_groups_count++;
+  }
+  LOG_GENERAL(INFO, "DEBUG num of shard group count " << shard_groups_count)
+
+  my_cluster_num = indexB2 / MULTICAST_CLUSTER_SIZE;
+  my_shards_lo = my_cluster_num * shard_groups_count;
+  my_shards_hi = my_shards_lo + shard_groups_count - 1;
+
+  if (my_shards_hi >= shards.size()) {
+    my_shards_hi = shards.size() - 1;
+  }
+}
+
+bool DataSender::SendDataToOthers(
+    const BlockBase& blockwcosig,
+    const deque<pair<PubKey, Peer>>& sendercommittee,
+    const DequeOfShard& shards, const VectorOfLookupNode& lookups,
+    const BlockHash& hashForRandom,
+    const ComposeMessageForSenderFunc& composeMessageForSenderFunc,
+    const SendDataToLookupFunc& sendDataToLookupFunc,
+    const SendDataToShardFunc& sendDataToShardFunc) {
+  if (LOOKUP_NODE_MODE) {
+    LOG_GENERAL(WARNING,
+                "DataSender::SendDataToOthers not expected "
+                "to be called from LookUp node.");
+    return true;
+  }
+
+  LOG_MARKER();
+
+  if (blockwcosig.GetB2().size() != sendercommittee.size()) {
+    LOG_GENERAL(WARNING, "B2 size and committee size is not identical!");
+    return false;
+  }
+
+  deque<pair<PubKey, Peer>> tmpCommittee;
+  for (unsigned int i = 0; i < blockwcosig.GetB2().size(); i++) {
+    if (blockwcosig.GetB2().at(i)) {
+      tmpCommittee.push_back(sendercommittee.at(i));
+    }
+  }
+
+  bool inB2 = false;
+  uint16_t indexB2 = 0;
+  for (const auto& entry : tmpCommittee) {
+    if (entry.second == Peer()) {
+      inB2 = true;
+      break;
+    }
+    indexB2++;
+  }
+
+  if (inB2) {
+    vector<unsigned char> message;
+    if (!(composeMessageForSenderFunc &&
+          composeMessageForSenderFunc(message))) {
+      LOG_GENERAL(
+          WARNING,
+          "composeMessageForSenderFunc undefined or cannot compose message");
+      return false;
+    }
+
+    uint16_t randomDigits =
+        DataConversion::charArrTo16Bits(hashForRandom.asBytes());
+    bool committeeTooSmall = tmpCommittee.size() < TX_SHARING_CLUSTER_SIZE;
+    uint16_t nodeToSendToLookUpLo =
+        committeeTooSmall
+            ? 0
+            : (randomDigits % (tmpCommittee.size() - TX_SHARING_CLUSTER_SIZE));
+    uint16_t nodeToSendToLookUpHi =
+        committeeTooSmall ? tmpCommittee.size()
+                          : nodeToSendToLookUpLo + TX_SHARING_CLUSTER_SIZE;
+
+    LOG_GENERAL(INFO, "lo: " << nodeToSendToLookUpLo << " hi: "
+                             << nodeToSendToLookUpHi << " my: " << indexB2);
+
+    if (indexB2 >= nodeToSendToLookUpLo && indexB2 < nodeToSendToLookUpHi) {
+      LOG_GENERAL(INFO,
+                  "Part of the committeement (assigned) that will send the "
+                  "Data to the lookup nodes");
+      if (sendDataToLookupFunc) {
+        sendDataToLookupFunc(lookups, message);
+      }
+    }
+
+    unsigned int my_cluster_num;
```

Comment:
Initialize variable when define it. Can change to unsigned int my_cluster_num = 0